### PR TITLE
Functions for gitlab

### LINKED ISSUE #125: GitLab support
It would be nice to finally start implementing the GitLab support.

We can probably combine the code from GitHub and the already present code.

The documentation can be found here: https://python-gitlab.readthedocs.io/en/stable/

### PR DESCRIPTION
**Pull request by lbarcziova**
Fixes #125 
Tests fail because test responses can't be saved.

### CODE REVIEW

File: ogr/services/gitlab.py [hunk 1]
```diff
@@ -41,13 +52,16 @@ class GitlabService(GitService):
...
+        self.g = gitlab.Gitlab(url=url, private_token=token, ssl_verify=False)
         self.g.auth()
-        self.user = self.g.users.list(username=self.g.user.get_username())[0]
         self.repo = None
         if full_repo_name:
             self.repo = self.g.projects.get(full_repo_name)
 
+    @property
+    def user(self) -> GitUser:
+        return self.g.users.list(username=self.g.user.get_username())[0]
```
Comment (lachmanfrantisek) [inline ogr/services/gitlab.py]: Similarily to GitHub/Pagure, we want to have `GitlabUser` class -- this property should return an instance of that class.

We need to forward a service instance (`self`) when creating that instance. Same as we did for GitHub:

https://github.com/packit-service/ogr/blob/master/ogr/services/github.py#L140
```python
    @property
    def user(self) -> GitUser:
        return GithubUser(service=self)
```
https://github.com/packit-service/ogr/blob/master/ogr/services/github.py#L801
```python
class GithubUser(BaseGitUser):
    service: GithubService

    def __init__(self, service: GithubService) -> None:
        super().__init__(service=service)

# :
# :

    def get_username(self) -> str:
        return self.service.github.get_user().login
```

File: ogr/services/gitlab.py [hunk 2]
```diff
@@ -112,23 +126,183 @@ def _fork_gracefully(target_repo):
...
+        return [commit.id for commit in mr.commits()]
+
+    @staticmethod
+    def _prcomment_from_gitlab_object(raw_comment) -> PRComment:
+        return PRComment(
+            comment=raw_comment.body,
+            author=raw_comment.author["username"],
+            created=raw_comment.created_at,
+            edited=raw_comment.updated_at,
+        )
```
Comment (lachmanfrantisek) [inline ogr/services/gitlab.py]: Please, move the static methods together (`***_from_gitlab_object` methods).

File: ogr/services/gitlab.py [hunk 3]
```diff
@@ -112,23 +126,183 @@ def _fork_gracefully(target_repo):
...
+        git_tag = self.repo.tags.get(tag_name)
+        return GitTag(name=git_tag.name, commit_sha=git_tag.commit["id"])
+
+    @staticmethod
+    def _release_from_gitlab_object(raw_release, git_tag: GitTag) -> Release:
+        return Release(
+            title=raw_release.name,
+            body=raw_release.description,
+            tag_name=raw_release.tag_name,
+            url=None,
```
Comment (lachmanfrantisek) [inline ogr/services/gitlab.py]: It looks like there is no such web-url. There is only a web-url for tags:

Example:
- releases: https://gitlab.gnome.org/World/OpenPaperwork/paperwork/-/releases
- tag: https://gitlab.gnome.org/World/OpenPaperwork/paperwork/-/tags/1.2.4

We can leave it empty or use the tag url.

File: ogr/services/gitlab.py [hunk 4]
```diff
@@ -41,13 +52,16 @@ class GitlabService(GitService):
...
         self.g.auth()
-        self.user = self.g.users.list(username=self.g.user.get_username())[0]
         self.repo = None
         if full_repo_name:
             self.repo = self.g.projects.get(full_repo_name)
 
+    @property
+    def user(self) -> GitUser:
+        return self.g.users.list(username=self.g.user.get_username())[0]
+
```
Comment (lachmanfrantisek) [inline ogr/services/gitlab.py]: To have the same API as GitHub/PAgure, we need to create a `GitlabProject` class.

- Feel free to remove unused methods form `GitlabService`. (e.g. `create_from_remote_url`, `is_fork_of`, ...) and move project-specific methods to newly created `GitlabProject`.

**Comment by lachmanfrantisek** [comment]
I've found the potential python-gitlab method that needs to be overwritten for mocking:

https://github.com/python-gitlab/python-gitlab/blob/master/gitlab/__init__.py#L442

I'll take a look tomorrow.

**Comment by lachmanfrantisek** [comment]
> I've found the potential python-gitlab method that needs to be overwritten for mocking:
> 
> https://github.com/python-gitlab/python-gitlab/blob/master/gitlab/__init__.py#L442
> 
> I'll take a look tomorrow.

I am just looking on it.

Feel free to work on the real implementation, and I'll send you another commit with the changes when I find the solution.

File: tests/integration/test_gitlab.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        )
+        PersistentObjectStorage().storage_file = persistent_data_file
+
+        if PersistentObjectStorage().is_write_mode and (
+            not self.user or not self.token
+        ):
+            raise EnvironmentError("please set GITLAB_TOKEN GITLAB_USER env variables")
+
+        self.service = GitlabService(
+            token=self.token, url="https://gitlab.cee.redhat.com"
```
Comment (lachmanfrantisek) [inline tests/integration/test_gitlab.py]: This is our internal instance. (It's not accessible from outside and we should not use internal links outside of our network...)

Please, use some easily-accessible instance (e.g. gitlab.com or gitlab.gnome.org).

**Comment by lachmanfrantisek** [comment]
@lbarcziova I've rebased and it looks like it works.

(We need to update url-mapping for gitlab. I'll send a follow-up PR with that.)

**Comment by lachmanfrantisek** [comment]
@lbarcziova if you know anything missing in the GitLab implementation, please, add it to the description of the [*Missing feature in GitLab implementation* issue](https://github.com/packit-service/ogr/pull/158).

**Comment by TomasTomecek** [comment]
aaaaaaa, can we get this to upsint, I hate touching the web ui

**Comment by lachmanfrantisek** [comment]
> aaaaaaa, can we get this to upsint, I hate touching the web ui

Some hackfest and redesign? I need that badly as well!

**Pull request merged**